plrust/src/allow_list.rs

```diff
@@ -212,6 +212,92 @@ pub enum Error {
+struct AllowedDependency {
+    name: String,
+    ver: String,
+    features: Vec<String>,
+    default_features: bool,
+}
+
+/// A container struct to wrap around a vector of AllowedDependency used by plrust.allowed_dependencies function
+pub struct AllowedDependencies {
+    allowed_dependencies: Vec<AllowedDependency>,
+}
+
+/// A type alias to wrap around the tuple signature of plrust.allowed_dependencies function.
+/// An AllowedDependency struct will be converted into this type.
+pub type AllowedDependencyTuple = (String, String, Vec<String>, bool);
+
+impl From<AllowedDependencies> for Vec<AllowedDependencyTuple> {
+    fn from(a: AllowedDependencies) -> Vec<AllowedDependencyTuple> {
+        a.allowed_dependencies
+            .iter()
+            .map(|item| {
+                (
+                    item.name.to_owned(),
+                    item.ver.to_owned(),
+                    item.features.to_owned(),
+                    item.default_features,
+                )
+            })
+            .collect()
+    }
+}
+
+/// Get all the allowed dependencies entries as a AllowedDependencies struct
+pub fn get_alllowed_dependencies() -> AllowedDependencies {
+    let allowlist: BTreeMap<String, Dependency> =
+        load_allowlist().expect("Error loading dependency allow-list");
+    let mut allowed_dependencies: Vec<AllowedDependency> = vec![];
+    for dependency in allowlist.values() {
+        let entries = get_entries_for_single_allowed_dependency(dependency);
+        allowed_dependencies.extend(entries)
+    }
+    AllowedDependencies {
+        allowed_dependencies,
+    }
+}
+
+/// Multiple versions of a dependency can be allowed. Create an AllowedDependency struct for each version of the dependency
+fn get_entries_for_single_allowed_dependency(dep: &Dependency) -> Vec<AllowedDependency> {
+    let mut entries: Vec<AllowedDependency> = vec![];
+    for version in dep.versions.values() {
+        let name = dep.name.clone();
+        let ver = version
+            .get("version")
+            .unwrap()
+            .to_string()
+            .replace("\"", "");
+        let features: Vec<String> = match version.get("features") {
+            Some(features) => match features.is_array() {
+                true => features
+                    .as_array()
+                    .unwrap()
+                    .iter()
+                    .map(|f| f.as_str().unwrap().to_owned())
+                    .collect(),
+                false => vec![features.as_str().unwrap().to_owned()],
+            },
+            None => vec![],
+        };
+        let default_features = match version.get("default-features") {
+            Some(default_features) => match default_features.is_bool() {
+                true => default_features.as_bool().unwrap(),
+                false => false,
+            },
+            None => true,
+        };
```

Comment:
Same concern as above, only one possible type is acceptable:

```suggestion
        let default_features = version
            .get("default-features")
            .map(|b| b.as_bool().unwrap())
            .unwrap_or(true);
```

Reply:
similar as above, the code that parses the file does not check the type of the value, so i assume `false` if the value is not a bool. Maybe that can be fixed separately.

Reply:
I have added some code to validate that default-features is a boolean